internal/controller/engine/engine.go

```diff
@@ -0,0 +1,434 @@
+	if err != nil {
+		return errors.Wrapf(err, "cannot stop watches for controller %q", name)
+	}
+
+	// Then stop any informers that only this controller was using.
+	if stopped > 0 {
+		if err := e.RemoveUnwatchedInformers(ctx); err != nil {
+			return errors.Wrapf(err, "cannot remove unwatched informers after stopping watches for controller %q", name)
+		}
+	}
+
+	// Finally, stop and delete the controller.
+	e.mx.Lock()
+	c.cancel()
+	delete(e.controllers, name)
+	e.mx.Unlock()
+
+	e.log.Debug("Stopped controller", "controller", name)
+	return nil
+}
+
+// IsRunning returns true if the named controller is running.
+func (e *ControllerEngine) IsRunning(name string) bool {
+	e.mx.RLock()
+	_, running := e.controllers[name]
+	e.mx.RUnlock()
+	return running
+}
+
+// Watch an object.
+type Watch struct {
+	kind       client.Object
+	handler    handler.EventHandler
+	predicates []predicate.Predicate
+}
+
+// WatchFor returns a Watch for the supplied kind of object. Events will be
+// handled by the supplied EventHandler, and may be filtered by the supplied
+// predicates.
+func WatchFor(kind client.Object, h handler.EventHandler, p ...predicate.Predicate) Watch {
+	return Watch{kind: kind, handler: h, predicates: p}
+}
+
+// StartWatches instructs the named controller to start the supplied watches.
+// The controller will only start a watch if it's not already watching the type
+// of object specified by the supplied Watch. StartWatches blocks other
+// operations on the same controller if and when it starts a watch.
+func (e *ControllerEngine) StartWatches(name string, ws ...Watch) error {
+	e.mx.RLock()
+	c, ok := e.controllers[name]
+	e.mx.RUnlock()
+
+	if !ok {
+		return errors.Errorf("controller %q is not running", name)
+	}
+
+	watchExists := make(map[schema.GroupVersionKind]bool)
+	c.mx.RLock()
+	for gvk := range c.sources {
+		watchExists[gvk] = true
+	}
+	c.mx.RUnlock()
+
+	// It's possible that we didn't explicitly stop a watch, but its backing
+	// informer was removed. This implicitly stops the watch by deleting its
+	// backing listener. If a watch exists but doesn't have an active informer,
+	// we want to restart the watch (and, implicitly, the informer).
+	a := e.infs.ActiveInformers()
+	activeInformer := make(map[schema.GroupVersionKind]bool, len(a))
+	for _, gvk := range a {
+		activeInformer[gvk] = true
+	}
+
+	// Using a map here deduplicates watches by GVK. If we're asked to start
+	// several watches for the same GVK at in the same call, we'll only start
+	// the last one.
+	start := make(map[schema.GroupVersionKind]Watch, len(ws))
+	for _, w := range ws {
+		gvk, err := apiutil.GVKForObject(w.kind, e.mgr.GetScheme())
+		if err != nil {
+			return errors.Wrapf(err, "cannot determine group, version, and kind for %T", w.kind)
+		}
+
+		// We've already created this watch and the informer backing it is still
+		// running. We don't need to create a new watch.
+		if watchExists[gvk] && activeInformer[gvk] {
```

Comment:
here we can lose a watch if `watchExists[gvk]` is true, but a parallel go routine removes the watch.

Reply:
Thanks!

Apparently I keep making this mistake. 😬 This same class of bug exists in the `PackagedFunctionRunner` locking code (https://github.com/crossplane/crossplane/issues/5661) and when I went to refresh my memory on the correct use of `RWLock` I realized I'd already read and forgotten about the various Go issues I found.